internal/pkg/client/request.go

```diff
@@ -31,28 +32,29 @@ type ResponseListener struct {
 type Request struct {
 	*resty.Request
 	*Response
-	lock            *sync.Mutex
-	id              int
-	sent            bool
-	done            bool
-	url             string
-	pathParams      map[string]string // for logs
-	queryParams     map[string]string // for logs
-	sender          Sender
-	cancelListeners context.CancelFunc  // to cancel execution of listeners if a new sub-request is added via WaitFor()
-	listeners       []*ResponseListener // callback invoked when request is completed
-	waitingFor      []*Request          // defer execution listeners until another request is completed
+	lock          *sync.Mutex
+	id            int
+	sent          bool
+	done          bool
+	url           string
+	pathParams    map[string]string // for logs
+	queryParams   map[string]string // for logs
+	sender        Sender
+	listeners     []*ResponseListener // callback invoked when request is completed
+	waitingFor    []*Request          // defer execution listeners until another request is completed
+	waitingForGrp *sync.WaitGroup     // blocks the execution of listeners until all sub-requests are completed
 }
 
 func NewRequest(id int, sender Sender, request *resty.Request) *Request {
 	return &Request{
-		Request:     request,
-		lock:        &sync.Mutex{},
-		id:          id,
-		pathParams:  make(map[string]string),
-		queryParams: make(map[string]string),
-		url:         request.URL,
-		sender:      sender,
+		Request:       request,
+		lock:          &sync.Mutex{},
+		id:            id,
+		pathParams:    make(map[string]string),
+		queryParams:   make(map[string]string),
+		url:           request.URL,
+		sender:        sender,
+		waitingForGrp: &sync.WaitGroup{},
```

Comment:
Inicializacia `WaitGroup`, viac info: https://developpaper.com/examples-of-the-usage-of-sync-waitgroup-in-golang/